task-maker-async/src/store.rs

```diff
@@ -473,190 +240,169 @@ impl StoreService {
+        let mut service = self.service.lock().unwrap();
+        let handle_info = service.get_handle_info(&handle)?;
+        if let Some(hasher) = &mut handle_info.main_file_hasher {
+            if file == FileSetFile::MainFile {
+                hasher.update(&data);
+            }
+        }
+        let hash = handle_info.hash;
+        let file_set = service.get_fileset(hash).unwrap();
+        if file_set.kind() == FileSetKind::InputFile
+            && matches!(file, FileSetFile::AuxiliaryFile(_, _))
+        {
+            return Err(Error::InvalidFileForInput(file));
         }
+        file_set.append_to_file(&file, &data);
         Ok(())
     }
 
-    async fn refresh_file_set_lease(
+    async fn finalize_file_set(
         self,
         context: Context,
-        handle: FileSetHandle,
+        handle: FileSetWriteHandle,
     ) -> Result<(), Error> {
         let mut service = self.service.lock().unwrap();
-        service.validate_and_refresh(handle).map(|x| ())
+        let handle_info = service.get_handle_info(&handle)?;
+        if let Some(hasher) = handle_info.main_file_hasher.take() {
+            let hash = hasher.finalize();
+            if *hash.as_bytes() != handle_info.hash.data {
+                return Err(Error::InvalidHash(handle_info.hash.data, *hash.as_bytes()));
+            }
+        }
+        let hash = handle_info.hash;
+        let file_set = service.get_fileset(hash).unwrap();
+        file_set.mark_finalized();
+        service.handles.remove(&handle.id);
+        Ok(())
     }
 
-    async fn open_file(
-        self,
-        context: Context,
-        handle: FileSetHandle,
-        file: FileSetFile,
-    ) -> Result<FileHandle, Error> {
+    async fn drop_computation(self, context: Context, hash: FileSetHash) -> Result<(), Error> {
         let mut service = self.service.lock().unwrap();
-        service.validate_and_refresh(handle)?;
-        service.open_file(handle, file)
+        let data_hm = service
+            .file_sets
+            .get_mut(&hash.data)
+            .ok_or(Error::UnknownHash(hash))?;
+        if let Some(fs) = data_hm.get_mut(&hash.variant) {
+            if fs.is_finalized() {
+                return Ok(());
+            }
+            data_hm.remove(&hash.variant).unwrap();
+        }
+        Ok(())
     }
 
-    async fn append_chunk(
+    async fn wait_for_fileset(
         self,
         context: Context,
-        file: FileHandle,
-        data: Vec<u8>,
+        hash: FileSetHash,
+        wait_for: WaitFor,
     ) -> Result<(), Error> {
-        let mut service = self.service.lock().unwrap();
-        service.validate_and_refresh(file.file_set_handle)?;
-        if file.file_set_handle.mode != HandleMode::Write {
-            return Err(Error::AppendRead(file.file_set_handle.id, file.id));
-        }
-        service.append_to_file(file, data);
-        Ok(())
+        let res = if wait_for == WaitFor::Finalization {
```

Comment:
`match` here is better than `if` since you can list both variants and fail to compile if additional variants are added.